### PR TITLE
localstore/pin: refactor and add paging

### LINKED ISSUE #464: Delete all pins (20 at a time?)
# User story

## Story

As a file host I want to be able to delete ALL current pins as I load a new consistent set of files into the Swarm.

I'm currently doing "Get list of pinned chunks" followed by "Unpin chunk with given address" for each returned address (times the returned pinCounter).  But imagine my surprise when this doesn't delete all the pins?  Another "Get list of pinned chunks" returns still more (and different) pins.

It seems that "Get list of pinned chunks" only returns 20 pins per invocation, but this is not mentioned in the documentation for the Debug API.  There also does not seem to be a way to specify how many pins are desired.

Is there a way to query ALL the pins?  Or is there a context so that different sets of pins can be queried?

## Acceptance criteria

At least document "Get list of pinned chunks" to have a limit of 20 pins.  But hopefully also describe how to query all existing pins.

Or provide an (Debug?) API that deletes all current pins with full understanding of the ramifications of this request.

## Background

I have a script that adds lots (1,000s) of files to the swarm and automatically pins each address while also recording said address externally.  As this file set is refreshed, I want to delete all current pins without consulting the external source and before re-adding the new file set.

## Tasks

| Task | Assignee |  Done (1), or 0 |
| ------ |------------| -----|

### PR DESCRIPTION
**Pull request by acud**
This PR adds pagination to the pin api, as well as some minor cleanups and refactorings.
A default 100 entry response is returned, and a hard limit of 1000 entries per response is enforced.

fixes #464

### CODE REVIEW

File: pkg/localstore/pin_test.go [hunk 1]
```diff
@@ -12,114 +12,132 @@ import (
...
+	err = db.Set(context.Background(), storage.ModeSetPin, swarm.NewAddress(chunk.Address().Bytes()))
+	if err != nil {
+		t.Fatal(err)
+	}
+	pinCounter, err = db.PinCounter(swarm.NewAddress(chunk.Address().Bytes()))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if pinCounter != 2 {
+		t.Fatalf("want pin counter 1 but got %d", pinCounter)
```
Comment (zelig) [inline pkg/localstore/pin_test.go:78]: 1. -> 2

**Comment by ldeffenb** [comment]
Are there defaults for limit and offset?  My (possibly incorrect) reading looks like they'll default to zero which is fine (and one could argue expected) for offset, but defaulting the limit to 0 results in (I think) just a single pin being returned.

For backward compatibility it would seem that a default limit, if not specified, of the old 20 would be good?

**Comment by ldeffenb** [comment]
I just re-read and saw "A default 100 entry response", but I didn't see that in the code?  I confess, I might have missed it though.

**Comment by ldeffenb** [comment]
Ignore me.  I now see the 100 limit in the "add error" commit.  Didn't expect that comment on that commit.

**Comment by acud** [comment]
@ldeffenb there's also a short description in the PR ^^ :)

File: pkg/api/pin.go [hunk 2]
```diff
@@ -96,16 +104,41 @@ type listPinnedChunksResponse struct {
...
+			s.Logger.Debugf("list pins: parse limit: %v", err)
+			s.Logger.Errorf("list pins: bad limit")
+			jsonhttp.BadRequest(w, "bad limit")
+		}
+	}
+
+	pinnedChunks, err := s.Storer.PinnedChunks(r.Context(), offset, limit)
 	if err != nil {
-		s.Logger.Debugf("debug-api: pin chunk: listing pinned chunks: %v", err)
+		s.Logger.Debugf("pin chunk: list pinned: %v", err)
```
Comment (janos) [inline pkg/api/pin.go:131]: `pin chunk` -> `list pins`

File: pkg/api/pin.go [hunk 3]
```diff
@@ -131,13 +166,14 @@ func (s *server) getPinnedChunk(w http.ResponseWriter, r *http.Request) {
...
 
-	pinCounter, err := s.Storer.PinInfo(addr)
+	pinCounter, err := s.Storer.PinCounter(addr)
 	if err != nil {
 		if errors.Is(err, storage.ErrNotFound) {
 			jsonhttp.NotFound(w, nil)
 			return
 		}
-		s.Logger.Debugf("debug-api: pin chunk: listing pinned chunks: %v", err)
+		s.Logger.Debugf("pin chunk: get pin counter: %v", err)
```
Comment (janos) [inline pkg/api/pin.go:175]: `pin chunk` -> `pin counter`

**Review by janos — APPROVED** [review]

**Comment by ldeffenb** [comment]
> @ldeffenb there's also a short description in the PR ^^ :)

I read that, but was checking the code just to see if I could see how it worked.

File: pkg/localstore/pin.go [hunk 4]
```diff
@@ -37,36 +32,34 @@ func (db *DB) PinnedChunks(ctx context.Context, cursor swarm.Address) (pinnedChu
...
+		if limit == 0 {
 			return true, nil
-		} else {
-			return false, nil
 		}
-
+		return false, nil
 	}, &shed.IterateOptions{
-		StartFrom:         &it,
 		SkipStartFromItem: false,
```
Comment (zbiljic) [inline pkg/localstore/pin.go:55]: No need for `SkipStartFromItem` option if the `StartFrom` is removed.
Options here can be replaced by `nil`.

**Review by zbiljic — APPROVED** [review]

**Pull request merged**